### PR TITLE
model: Add footer notification for messages.

### LINKED ISSUE #781: Notify user if sent/edited message is outside current narrow
For example, if a user is in topic narrow and edits a message's topic then it would be moved do a different narrow. We should probably display a message in the footer if this happens.

### PR DESCRIPTION
**Pull request by zee-bit**
This will set a footer text notifying the user everytime
he sends or edits a message that is outside the current
narrow. This will give the user a pointer as to why the
message disappeared from the screen.

Fixes #781.

### CODE REVIEW

**Review by neiljp — COMMENTED** [review]
@zee-bit Thanks for giving this issue a go!

I just tried this out and it doesn't seem to work quite as I'd expect, in that it's not triggered in at least this example case:
* narrow to test here > zt1
* compose and send a message to test here > zt2

Note that this shows a message like we want to see, in the webapp.

It may be worth checking the webapp logic to confirm sending from/to all-messages/PMs/mentioned/streams/topics narrows.

You've left quite a lot of debugging statements in there - unless it's relevant to discussion points, it's better to leave those out in the version you push to a PR - you can always keep it in a branch locally and add it back in on a local branch, in a separate commit.

File: zulipterminal/model.py [hunk 1]
```diff
@@ -1011,11 +1018,21 @@ def _handle_update_message_event(self, event: Event) -> None:
...
                 for msg_id in event['message_ids']:
                     self.index['messages'][msg_id]['subject'] = new_subject
                     self._update_rendered_view(msg_id)
 
+                view = self.controller.view
+                print("event=>", event, flush=True)
+                if (len(self.narrow) == 2
+                    and self.narrow[1][0] == 'topic'
+                        and old_subject != new_subject):
+                    view.set_footer_text("Subject changed", 3)
```
[the remark this answers is not in the document]
Comment (zee-bit) [reply zulipterminal/model.py:1034]: No, it's for the same issue. This was for the case when a message's subject is edited outside of the current narrow. So the message disappears from the current narrow and is shifted to some other narrow. But, I think the logic behind this is also incorrect. I'll have to think more on this. Any pointers that you would like to give?

[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/model.py:1034]: This is sort of the same issue, but I think it would be fine to first work on showing a message when sending to a different narrow than the current one, then work on what happens with editing as a follow-up.

I think in both cases the focus is on what the active user does within this ZT session, so it's likely best to hook into the messages to the server, rather than events from it.

File: zulipterminal/model.py [hunk 2]
```diff
@@ -966,6 +968,10 @@ def _handle_message_event(self, event: Event) -> None:
                              and self.narrow[1][1] == message['subject']))):
                     msg_log.append(msg_w)
 
+                if not append_to_stream:
+                    self.controller.view.set_footer_text(
+                        'Message sent outside of current narrow', 3)
```
[the remark this answers is not in the document]
Comment (zee-bit) [reply zulipterminal/model.py]: I have updated the code, to fix the above issue as well as extended it for footer notification outside of streams. I have also amended some tests for `model` and now `make check` is working fine.

File: zulipterminal/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zee-bit) [reply zulipterminal/model.py]: I have moved the code for footer notification in case of a message update, to hook to the part where the request to the server is sent. Can you please review my PR whenever you get the time?

**Review by neiljp — COMMENTED** [review]
@zee-bit This is definitely in the right area of the code now, but note that narrows are in general much more complex than this - including, for example, if I write a PM and I'm in a stream narrow (or vice versa).

Rather than spreading this logic around, I'd suggest making a method which you can test independently and then use in all the locations where we send/update. This could be useful for reasoning about other cases too.

The other aspect you've not considered here is error-handling - we shouldn't show anything if the message is not sent/updated successfully.

**Comment by zee-bit** [comment]
@zulipbot add "PR needs review"
@zulipbot remove "PR needs update"

File: zulipterminal/helper.py [hunk 3]
```diff
@@ -651,6 +651,28 @@ def display_error_if_present(response: Dict[str, Any], controller: Any
...
+            controller.view.set_footer_text(
+                'Message is sent outside of current narrow.', 3)
+    elif request['type'] == 'private':
+        pm_narrow = [['is', 'private']]
+        pm_with_narrow = [['pm_with', request['to']]]
+
+        if (current_narrow != pm_narrow
+                and current_narrow != pm_with_narrow):
+            controller.view.set_footer_text(
+                'Message is sent outside of current narrow.', 3)
```
Comment (neiljp) [inline zulipterminal/helper.py:673]: This is duplicated, which suggests to me this could be simpler.

File: zulipterminal/model.py [hunk 4]
```diff
@@ -423,6 +429,13 @@ def update_stream_message(self, topic: str, message_id: int,
 
         response = self.client.update_message(request)
         display_error_if_present(response, self.controller)
+        if response['result'] == 'success':
+            old_topic = self.index['messages'][message_id].get('subject', None)
+            new_topic = request['topic']
+            view = self.controller.view
+            if old_topic != new_topic:
+                view.set_footer_text('Topic changed', 3)
```
Comment (neiljp) [inline zulipterminal/model.py:437]: Could this text be expanded to be more descriptive? There's no test for this.

File: zulipterminal/model.py [hunk 5]
```diff
@@ -385,6 +386,9 @@ def send_private_message(self, recipients: List[str],
             }
             response = self.client.send_message(request)
             display_error_if_present(response, self.controller)
+            # FIXME: Is there a better way?
```
Comment (neiljp) [inline zulipterminal/model.py:389]: To do what? You might combine the boolean test into a variable in any case.

**Review by neiljp — COMMENTED** [review]
@zee-bit This seems to work mostly fine, though there are edge cases in terms of narrows that behave unexpectedly, which you might clarify with tests. I'm also wondering if a longer footer duration may be wise, though this may be simplified if we can use footer message 'themes'?

As per my message in zulip-terminal, #924 may be relevant, though don't feel a need to base this PR on it.

File: zulipterminal/helper.py [hunk 6]
```diff
@@ -651,6 +651,28 @@ def display_error_if_present(response: Dict[str, Any], controller: Any
         controller.view.set_footer_text(response['msg'], 3)
 
 
+def notify_if_message_sent_outside_narrow(request: Dict[str, Any],
```
[the remark this answers is not in the document]
Comment (zee-bit) [reply zulipterminal/helper.py:654]: How about `message`?

[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/helper.py:654]: That's good, though the type will presumably be a Composition now?

[the remark this answers is not in the document]
Comment (zee-bit) [reply zulipterminal/helper.py:654]: For that, we would need to type-annotate our `request` dict(in `model`) to appropriate `Composition`?

I tried doing this, but it seems like MyPy doesn't allow converting existing dict to `TypedDict`(see [mypy issue #8890](https://github.com/python/mypy/issues/8890)). Is there any other way of doing so?

File: tests/helper/test_helper.py [hunk 7]
```diff
@@ -309,6 +310,31 @@ def test_display_error_if_present(mocker, response, footer_updated):
...
 
+@pytest.mark.parametrize(['req', 'narrow', 'footer_updated'], [
+    ({'type': 'private', 'to': 'foo@gmail.com', 'content': 'bar'},
+     [['is', 'private']], False),
+    ({'type': 'private', 'to': 'bar@chat.zulip.com', 'content': 'foo'},
+     [['pm_with', 'foo@chat.zulip.com']], True),
+    ({'type': 'stream', 'to': 'ZT', 'subject': '1', 'content': 'foo'},
+     [['stream', 'ZT'], ['topic', '1']], False),
+    ({'type': 'stream', 'to': 'here', 'subject': 'pytest', 'content': 'py'},
+     [['stream', 'test here']], True)
```
[the remark this answers is not in the document]
Comment (zee-bit) [reply tests/helper/test_helper.py:321]: Added. :)

**Comment by zee-bit** [comment]
@zulipbot add "PR needs review"
@zulipbot remove "PR needs update"

**Comment by neiljp** [comment]
@zee-bit Merged manually as 3260e3a1d004cf9462873f92fdbe2b19ae9cd6e1 after making a few minor adjustments :tada:
(as agreed in the stream)
